### PR TITLE
Add the ability to override download address

### LINKED ISSUE #23: Downloading to arbitrary addresses
A previous version of the crate had the ability to override the address, which was removed [here](https://github.com/dfu-rs/dfu-core/pull/20/files#diff-44e47ea9be0c53791e7aed4ba12a105bc663405ad3dec0008bc90c85505e1e11L40-L44). I was wondering if some version of it could be reintroduced to support devices for which we cannot write the firmware to the address returned by the descriptor, for whatever reason.

The device I am targeting returns this memory layout:

```
@Internal Flash   /0x08000000/16*01Ka,1*47Kg,1*1Kg
```

The first 16KB are read-only, because that's where the DFU bootloader itself is located. So I have to download the upgrade to 0x08004000 instead of 0x08000000. If I try to upgrade it with the crate as is, it just gets stuck in an infinite loop.

I don't know if the most "correct" solution for this would be to just set the address manually or have the crate skip the non-writable sectors when determining the start address, or to have the device itself omit the non-writable part when sending the descriptor, but I am putting it out there in case someone wants to discuss it.

I suppose setting arbitrary addresses would also allow using .hex files, which already include the address for each chunk of data, so this crate could be used in conjunction with the [ihex](https://crates.io/crates/ihex) crate. But I've never tried that and I don't know if it's that straightforward.

### PR DESCRIPTION
**Pull request by YgorSouza**
Closes #23

### CODE REVIEW

**Comment by YgorSouza** [comment]
This implementation seems to get the job done for my use case, but I'll see if I can improve it, as mentioned in the commit message.

In particular, I am not sure if this is applicable to the regular DFU, as the address seems to be hardcoded to 0 for it here:

https://github.com/dfu-rs/dfu-core/blob/fc9dee3e03c1f5133e8f99a42dd5135407f98f57/src/download.rs#L30-L32

And the address seems to only be set for the DfuSe anyway:

https://github.com/dfu-rs/dfu-core/blob/fc9dee3e03c1f5133e8f99a42dd5135407f98f57/src/download.rs#L106-L115

I don't have a non-DfuSe device to test, and I don't understand the protocol well enough, so I left it alone for now. But if anyone has ideas about that, it would be appreciated.

**Comment by cecton** [comment]
> I don't have a non-DfuSe device to test, and I don't understand the protocol well enough, so I left it alone for now. But if anyone has ideas about that, it would be appreciated.

hey @sjoerdsimons  :wave: if you could provide your input on this it would be really appreciated. Otherwise it's okay I'm fine with the safe approach currently implemented.

File: src/lib.rs [hunk 1]
```diff
@@ -218,6 +225,11 @@ impl<IO: DfuIo> DfuSansIo<IO> {
         Ok(())
     }
 
+    /// Set the address onto which to download the firmware
```
Comment (cecton) [inline src/lib.rs:228]: (just for consistency)
```suggestion
    /// Set the address onto which to download the firmware.
```

File: src/sync.rs [hunk 2]
```diff
@@ -69,6 +69,14 @@ where
         }
     }
 
+    /// Override the address onto which the firmware is downloaded.
+    ///
+    /// This address is only used if the device uses the DfuSe protocol.
+    pub fn with_override_address(&mut self, address: u32) -> &mut Self {
```
Comment (cecton) [inline src/sync.rs:75]: The original was called just "override_address". tbh I'm not sure which one is better. Any opinion on this?

Comment (YgorSouza) [reply src/sync.rs:75]: I put `with_` to be consistent with `with_progress` below, but removing it makes it more consistent with the std library, including the [example](https://rust-lang.github.io/api-guidelines/type-safety.html#non-consuming-builders-preferred) from the guidelines, so I'll change it.

**Comment by cecton** [comment]
If you don't mind I will wait a day or two to see if I can get more feedback. Don't hesitate to ping me if you see no more activity, I tend to forget things easily :sweat_smile:

**Comment by cecton** [comment]
Oh wow that has waited for way too long

**Comment by davidjmeyer** [comment]
Oh this is fantastic, I can use this library to successfully program a custom bootloader on my chip, and this was this missing piece!

**Comment by cecton** [comment]
I just made the release now

**Pull request merged**